text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,267 @@
+for non-v2 AWS-vended libraries such as the `@aws-solutions-constructs` libraries.
+
+Class name | File name | Desired identifier
+--|--|--
+Cluster | .../node_modules/aws-cdk-lib/lib/aws-ecs/lib/cluster.js | aws-cdk-lib.aws_ecs.Cluster@2.1.0
+Cluster | .../node_modules/aws-cdk-lib/lib/aws-eks/lib/cluster.js | aws-cdk-lib.aws_eks.Cluster@2.1.0
+LambdaToSqs | .../node_modules/@aws-solutions-constructs/aws-lambda-sqs/lib/index.js | @aws-solutions-constructs/aws-lambda-sqs.LambdaToSqs@1.63.0
+
+It would be nice if we used submodule naming that's consistent with the jsii
+naming. There is nothing in the jsii manifest that could reliably be used for
+this, but we can use an additional lookup table that can be generated by the
+monocdk/v2 build tool, as that is in charge of the submodule names anyway.
+
+There will be a `submodules.json` next to `package.json` with the following structure:
+
+```
+{
+  "lib/aws-ecs": "aws_ecs",
+  "lib/aws-eks": "aws_eks",
+  ...
+}
+```
+
+Which is used to map subdirectories to submodule names. Absence of this file indicates no submodule names
+need to be inserted.
+
+The first matching directory found will be used, so should we ever need a
+situation to support submodules-in-submodules, we can make use of the fact
+that JSON dictionaries are ordered and make sure the deeper directories occur
+first.
+
+> Implementation note: looking up the source file from a class is a potentially
+> expensive operation as it takes a linear scan through all sources for every class.
+> We should build a reverse index in order to speed this up.
+
+## Metadata resource encoding
+
+A rough estimate of the maximum number of constructs used in a stack will be
+about 400 (maximum of 200 L1s with a corresponding L2 wrapper). We will
+therefore need to encode roughly 400 strings of the form
+`@aws-solutions-constructs/aws-lambda-sqs.LambdaToSqs@1.63.0`.
+
+The longest overall type in the current monocdk is
+`monocdk-experiment.aws_cognito.CfnUserPoolRiskConfigurationAttachment.CompromisedCredentialsRiskConfigurationTypeProperty`
+(121 characters), the longest construct class name is
+`monocdk-experiment.aws_ecs_patterns.ApplicationMultipleTargetGroupsFargateService`
+(81 characters).
+
+If an identifier is estimated to be a maximum of 100 bytes (which seems
+reasonable be plenty of room) the resulting data set will be ~40kB.
+
+We can encode the elements as a list straight in the template, but it will
+take up a lot of space in the template: vertical space while looking at it,
+as well as bytes that count toward the template size (maximum of 460kB).
+
+Experiments with a random sampling of 400 resources from the list of classes:
+
+| Description | Raw size | Base64ed size |
+|--|--|
+| Plaintext | 23k | - |
+| gzipped | 3.6k | 4.8k |
+| Zopfli | 3.4k | 4.6k |
+| Bloom Filter (1% error) | 720b | 960b |
+| Tree-encoded (plaintext) | 7.4k | - |
+| Tree-encoded gzipped | 3.0k | 4.0k |
```

Comment:
4-5K seems like reasonable amount of metadata to be added. What does this look like in terms of percentage of the template's size?

Are you happy with these numbers?